### ISSUE TITLE
Abnormal behavior of TEST Metadata :Correct Label names were detected on Trained images set but Incorrect labels  were detected on test images .

### ISSUE THREAD

**Issue by ajbot2019** [facebookresearch/detectron2#942]
## Instructions To Reproduce the Issue:
I followed the instruction shared in colab notebook in the git repo of detectron2 .

Before training i verified if my train/test image and its annotations file were as per the exception of  detectron  format  using the code below and visualized the same and the classes/labels were detected as expected.

import random
from detectron2.utils.visualizer import Visualizer

for d in random.sample(traindataset, 3):
    img = cv2.imread(d["file_name"])
    visualizer = Visualizer(img[:, :, ::-1], metadata=train_metadata, scale=0.5)
    vis = visualizer.draw_dataset_dict(d)
    cv2_imshow(vis.get_image()[:, :, ::-1])
 
 PROBLEM:
After training i tried to detect with the newly trained weights for my custom dataset that had 3 labels /classes (hemlet,sunglass,shoe)

1.when tested the newly trained model  on the   images which was part of the training set itself the model  detected the regions and labelled it with correct labels names .

```
path_img='/content/train1.png'     
im = cv2.imread(path_img)
outputs = predictor(im)
print(outputs["instances"])
v = Visualizer(im[:, :, ::-1],
                metadata=train_metadata
                scale=0.5, 
                instance_mode=ColorMode.IMAGE_BW   # remove the colors of unsegmented pixels
)
v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
cv2_imshow(v.get_image()[:, :, ::-1])
```

**BUT when tested the newly trained model on images in   test set which were  was not part of the training   the 3 regions in test images were detected but it has incorrect label names on them .(for example shoe detected with bounding box but was labelled as  "helmet" )**

```
path_img='/content/test1.png'     
im = cv2.imread(path_img)
outputs = predictor(im)
v = Visualizer(im[:, :, ::-1],
                metadata=test_metadata
                scale=0.5, 
                instance_mode=ColorMode.IMAGE_BW   # remove the colors of unsegmented pixels
)
v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
cv2_imshow(v.get_image()[:, :, ::-1])
```

Not sure why the model was not able to detected correct labels on images that were not part of training  i used metadata=test_metadata while trying to predict on test images.

The abnormal behavior :
when i changed the metadata to train_metadata it was detecting the regions and labelled it with the expected label name .

```
path_img='/content/test1.png'     
im = cv2.imread(path_img)
outputs = predictor(im)
v = Visualizer(im[:, :, ::-1],
                metadata=train_metadata
                scale=0.5, 
                instance_mode=ColorMode.IMAGE_BW   # remove the colors of unsegmented pixels
)
v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
cv2_imshow(v.get_image()[:, :, ::-1])
```

1.Should we use the metadata of training set for testing on new images ? or metadata of testing set for testing on new images ?  

Below are the config details for training, Registration of train/test dataset and json file were done  as instructed in colab notebook:

```
> cfg = get_cfg()
> cfg.merge_from_file(model_zoo.get_config_file("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml"))
> cfg.DATASETS.TRAIN = ("my_dataset",)
> cfg.DATASETS.TEST = ()
> cfg.DATALOADER.NUM_WORKERS = 2
> cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml")  # Let training initialize from model zoo
> cfg.SOLVER.IMS_PER_BATCH = 2
> cfg.SOLVER.BASE_LR = 0.00025  # pick a good LR
> cfg.SOLVER.MAX_ITER = 40000     
> cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset (default:   )
> cfg.MODEL.ROI_HEADS.NUM_CLASSES = 3  # only has one class (ballon)
> 
```

I changed only values of  the iterations and classes and others were let default .
After training In tensorborad   the mask_rccn section  graph showed these observations , the false negative and false positive curve started at high initially   and  came down in final iterations .
Accuracy showed 0.95

any help on what would be the issue ,should i train for longer iterations ?
Any suggestions on how much iterations we should keep for training images with 10 or more classes ? and should we adjust learning rate in config files at different iterations ?
There weren't much documentations on these best practices of the values that should be set in config file for training custom dataset for detecting 10 or more classes request some help on them .

**Comment by ppwwyyxx** [comment]
As the [issue template](https://github.com/facebookresearch/detectron2/blob/master/.github/ISSUE_TEMPLATE/unexpected-problems-bugs.md) mentions:

> If you expect the model to converge / work better, note that we do not give suggestions on how to train a new model. Only in one of the two conditions we will help with it: (1) You're unable to reproduce the results in detectron2 model zoo. (2) It indicates a detectron2 bug.

We provide configs & models with standard academic settings and expect users to have the knowledge to choose or design appropriate models & parameters for their own tasks.
